### PR TITLE
Fix exists() for empty environment variable

### LINKED ISSUE #3266: exists('$VAR') returns 0 for existing, but empty variable
Some programs behave differently with empty, but existing variable and such exists() behaviour + #3265 makes it impossible to save and restore environment variables correctly. This issue was hit with `$TZ` variable, in Vim:

```
let saved_tz = $TZ
let $TZ = 'UTC'
echo strftime('%H', 0)
let $TZ = saved_tz
echo strftime('%H', 0)
```

will echo `00` two times instead of `00` and `03`. (In NeoVim it will always echo 03 because of #3264, but this code will change behaviour of `system('date')`.)

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/neovim/neovim/issues/3266.

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -8684,7 +8684,7 @@ static void f_exists(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   const char *p = tv_get_string(&argvars[0]);
   if (*p == '$') {  // Environment variable.
     // First try "normal" environment variables (fast).
-    if (os_getenv(p + 1) != NULL) {
+    if (os_env_exists(p + 1)) {
```
Comment (justinmk) [inline src/nvim/eval.c]: in `os_getenv` docstring , let's add:

    /// @see os_env_exists

Comment (blueyed) [reply src/nvim/eval.c]: Will do that in https://github.com/neovim/neovim/pull/10655 - or as a followup to https://github.com/neovim/neovim/pull/2574#issuecomment-516342975.

**Comment by blueyed** [comment]
Interestingly Neovim is not affected by `let $FOO = ''` unsetting the env on Windows (as seen by Vim at least): https://github.com/vim/vim/pull/4750#issuecomment-516357555

**Comment by blueyed** [comment]
According to https://github.com/vim/vim/pull/4750#issuecomment-516369883 an empty environment variable actually is more natural to be unset on Windows.

I wonder if we should follow that route?

**Comment by justinmk** [comment]
> According to [vim/vim#4750 (comment)](https://github.com/vim/vim/pull/4750#issuecomment-516369883) an empty environment variable actually is more natural to be unset on Windows.
> 
> I wonder if we should follow that route?

"Natural" there means "people are used to this weird behavior". And it means that exists() behaves differently on Windows vs Unix. How is that useful?

**Comment by blueyed** [comment]
Ok.  I'd say to wait a bit then how / if the patch for the test is handled in Vim, and then amend it accordingly, e.g. by stating that this is saner in Neovim then with the test.

File: src/nvim/testdir/test_exists.vim [hunk 2]
```diff
@@ -70,6 +70,9 @@ func Test_exists()
   call assert_equal(1, exists('$EDITOR_NAME'))
   " Non-existing environment variable
   call assert_equal(0, exists('$NON_ENV_VAR'))
+  " Empty environment variable
+  let $EDITOR_NAME = ''
+  call assert_equal(1, exists('$EDITOR_NAME'))
```
Comment (blueyed) [inline src/nvim/testdir/test_exists.vim]: So Vim is unlikely to fix it, and maybe not even taking (a different) test here.
Better to have this as a busted test then I assume?

File: test/functional/eval/exists_spec.lua [hunk 3]
```diff
@@ -0,0 +1,12 @@
+local helpers = require('test.functional.helpers')(after_each)
```
Comment (justinmk) [inline test/functional/eval/exists_spec.lua]: I moved this into `eval/environ_spec.lua`. Our test files have too much granularity, and really the common theme for these tests is "environment", not the particular VimL function. (Maybe should live in `core/environ_spec.lua` instead.)

**Comment by justinmk** [comment]
Oh, I did not realize this PR fails on AppVeyor.

```
  [  FAILED  ] test/functional\eval\exists_spec.lua @ 7: exists() handles empty env variable
  test/functional\eval\exists_spec.lua:9: Expected objects to be the same.
  Passed in:
  (number) 0
  Expected:
  (number) 1
```

**Pull request closed without merging**